fn.go

```diff
@@ -24,15 +38,227 @@ func (f *Function) RunFunction(_ context.Context, req *fnv1beta1.RunFunctionRequ
 
 	rsp := response.To(req, response.DefaultTTL)
 
+	// Get function input.
 	in := &v1beta1.Input{}
 	if err := request.GetInput(req, in); err != nil {
 		response.Fatal(rsp, errors.Wrapf(err, "cannot get Function input from %T", req))
 		return rsp, nil
 	}
 
-	// TODO: Add your Function logic here!
-	response.Normalf(rsp, "I was run with input %q!", in.Example)
-	f.log.Info("I was run!", "input", in.Example)
+	// Get XR the pipeline targets.
+	oxr, err := request.GetObservedCompositeResource(req)
+	if err != nil {
+		response.Fatal(rsp, errors.Wrapf(err, "cannot get observed composite resource"))
+		return rsp, nil
+	}
+	f.log.Debug("asdf", "bsdf", oxr)
+
+	// Build extraResource Requests.
+	requirements, err := buildRequirements(in, oxr)
+	if err != nil {
+		response.Fatal(rsp, errors.Wrapf(err, "could not build extraResource requirements"))
+		return rsp, nil
+	}
+	rsp.Requirements = requirements
+
+	// Verify that extra resources were even requested in input.
+	// Appears 100% needed.
+	// Still not sure why. This condition should only happen on a bad request.
+	if req.ExtraResources == nil {
+		f.log.Debug("No extra resources specified, exiting", "requirements", rsp.GetRequirements())
+		return rsp, nil
+	}
+
+	// Actually fetch extra resources vie gRPC to Crossplane.
+	extraResources, err := request.GetExtraResources(req)
+	if err != nil {
+		response.Fatal(rsp, errors.Wrapf(err, "fetching extra resources %T", req))
+		return rsp, nil
+	}
+
+	// Sort and verify min/max selected.
+	// Sorting is required for determinism.
+	verifiedExtras, err := verifyAndSortExtras(in, extraResources)
+	if err != nil {
+		return nil, errors.Wrapf(err, "sorting and verifying results")
+	}
+
+	// For now cheaply convert to JSON for serializing.
+	// TODO: look into resources.AsStruct or simlar since unsturctured k8s objects are already almost json.
```

Comment:
I totally saw that these k8s unstructured.Unstructured objects can be directly serialized, but I don't know how to do that while they're in a `map[string][]`.

I also saw the recommended AsStruct method skips that and goes straight to JSON in some cases anyway.

Reply:
🤔 this will do for now, we can think of improving it later on 👍